### PR TITLE
@impossible/non whitelist mint if

### PR DESCRIPTION
**Pull request by Kalzak**

### CODE REVIEW

File: Impossible-Agent-Suite/non-whitelist-mint-if/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  - Severity is always set to "high"
+  - Type is always set to "suspicious"
+  - Finding metadata
+    - `receiver`: The non-whitelisted account that received the tokens from the mint
+
+## Test Data
+
+The agent behaviour can be verified with the following transactions (Ethereum Network):
+
+- `0xe8d2ee70539e9dae26704569d6576ba0f5137908347a42a482fa22e8f88217bf` (IF - Mint to authorized address)I
```
Comment (yasmine9) [inline Impossible-Agent-Suite/non-whitelist-mint-if/README.md:32]:
```suggestion
- `0xe8d2ee70539e9dae26704569d6576ba0f5137908347a42a482fa22e8f88217bf` (IF - Mint to authorized address)
```

File: Impossible-Agent-Suite/non-whitelist-mint-if/package.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    "enable": "forta-agent enable",
+    "keyfile": "forta-agent keyfile",
+    "test": "jest"
+  },
+  "dependencies": {
+    "forta-agent": "^0.0.36"
+  },
+  "devDependencies": {
+    "forta-agent-tools": "^1.0.34",
+    "bignumber.js": "^9.0.1",
```
[the remark this answers is not in the document]
Comment (Kalzak) [reply Impossible-Agent-Suite/non-whitelist-mint-if/package.json]: `bignumber.js` dependency has been removed.

File: Impossible-Agent-Suite/non-whitelist-mint-if/src/agent.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+      // Get all `Transfer` events emitted from `tokenAddress`
+      const transfers = tx.filterLog(ERC20_ABI, tokenAddress);
+      // For each transfer
+      await Promise.all(transfers.map(async (transfer) => {
+        // If the `from` address is the zero address
+        if(transfer.args.from == ZERO_ADDR) {
+          // This transfer is a mint so run the verifier on the recipient address
+          const whitelisted = await verifierFunction(transfer.args.from);
+
+          console.log('whitelisted: ' + whitelisted);
```
[the remark this answers is not in the document]
Comment (Kalzak) [reply Impossible-Agent-Suite/non-whitelist-mint-if/src/agent.ts]: Test use of `console.log` has been removed.

File: Impossible-Agent-Suite/non-whitelist-mint-if/src/agent.spec.ts [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        'IF',
+        'IMPOSSIBLE-2-1'
+      ],
+      [
+        IDIA_ADDR,
+        mockedVerifier_IDIA,
+        'IDIA',
+        'IMPOSSBILE-2-2'
+      ]
+    ];
```
[the remark this answers is not in the document]
Comment (Kalzak) [reply Impossible-Agent-Suite/non-whitelist-mint-if/src/agent.spec.ts]: Mock function setup has been moved to `beforeAll` and the mock function reset has been placed in `beforeEach`.

File: Impossible-Agent-Suite/non-whitelist-mint-if/README.md (hunk 1 above)
Comment (Kalzak) [reply Impossible-Agent-Suite/non-whitelist-mint-if/README.md]: Typo fixed.

File: Impossible-Agent-Suite/non-whitelist-mint-if/src/agent.spec.ts [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+} from 'forta-agent-tools';
+
+import {
+  ERC20_ABI,
+  provideHandleTransaction
+} from './agent';
+
+import {
+  ethers
+} from 'ethers';
```
[the remark this answers is not in the document]
Comment (Kalzak) [reply Impossible-Agent-Suite/non-whitelist-mint-if/src/agent.spec.ts]: Specific package import has been done for `BigNumber` but having issues with `Interface`.

File: Impossible-Agent-Suite/non-whitelist-mint/src/agent.ts [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+  Verifier,
+  IF_WHITELIST_VERIFIER,
+  IDIA_WHITELIST_VERIFIER,
+} from './verifier';
+
+// Address of the Impossible Finance token contract
+let IF_ADDR = '0xb0e1fc65c1a741b4662b813eb787d369b8614af1';
+let IDIA_ADDR = '0x0b15ddf19d47e6a86a56148fb4afffc6929bcb89';
+let ZERO_ADDR = '0x0000000000000000000000000000000000000000';
+
```
Comment (yasmine9) [inline Impossible-Agent-Suite/non-whitelist-mint/src/agent.ts:19]:
```suggestion
// Address of the Impossible Finance token contract
const IF_ADDR = '0xb0e1fc65c1a741b4662b813eb787d369b8614af1';
const IDIA_ADDR = '0x0b15ddf19d47e6a86a56148fb4afffc6929bcb89';
const ZERO_ADDR = '0x0000000000000000000000000000000000000000';

```

File: Impossible-Agent-Suite/non-whitelist-mint/src/agent.ts [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    IDIA_WHITELIST_VERIFIER,
+    'IDIA',
+    'IMPOSSBILE-2-2'
+  ]
+];
+
+// Impossible Finance token contract ABI with relevant events/functions
+export const IF_ABI= [
+  'event Transfer(address indexed from, address indexed to, uint256 value)',
+  'function staxMigrate(uint)'
```
Comment (yasmine9) [inline Impossible-Agent-Suite/non-whitelist-mint/src/agent.ts:39]: recheck function signature

File: Impossible-Agent-Suite/non-whitelist-mint/README.md [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+## Description
+
+This agent detects IF and IDIA token mints to non-whitelisted addresses, excluding mints through the function `staxMigrate`
+
+## Supported Chains
+
+- Ethereum
+- Binance Smart Chain
+
```
Comment (yasmine9) [inline Impossible-Agent-Suite/non-whitelist-mint/README.md:11]: based on docs, Polygon is a supported chain as well

**Review by yasmine9 — CHANGES_REQUESTED** [review]

File: Impossible-Agent-Suite/non-whitelist-mint/src/agent.ts (hunk 6 above)
Comment (Kalzak) [reply Impossible-Agent-Suite/non-whitelist-mint/src/agent.ts:19]: This has been changed in the latest commit.

File: Impossible-Agent-Suite/non-whitelist-mint/README.md (hunk 8 above)
Comment (Kalzak) [reply Impossible-Agent-Suite/non-whitelist-mint/README.md:11]: Same situation as [here](https://github.com/NethermindEth/Forta-Agents/pull/154#discussion_r811611087)

File: Impossible-Agent-Suite/non-whitelist-mint/src/agent.ts (hunk 7 above)
Comment (Kalzak) [reply Impossible-Agent-Suite/non-whitelist-mint/src/agent.ts:39]: Argument name has been added to the `staxMigrate` function signature.

**Review by yasmine9 — APPROVED** [review]

**Pull request merged**